scipy/stats/_qmc.py

```diff
@@ -0,0 +1,1493 @@
+"""Quasi-Monte Carlo engines and helpers."""
+from abc import ABC, abstractmethod
+import math
+import warnings
+
+import numpy as np
+
+from scipy.optimize import brute
+from scipy._lib._util import check_random_state
+from scipy.optimize import basinhopping
+from scipy.stats import norm
+from scipy.stats._sobol import (
+    initialize_v, _cscramble, _fill_p_cumulative, _draw, _fast_forward,
+    _categorize, initialize_direction_numbers, _MAXDIM, _MAXBIT
+)
+
+__all__ = ['scale', 'discrepancy', 'QMCEngine', 'Sobol', 'Halton',
+           'OrthogonalLatinHypercube', 'LatinHypercube', 'OptimalDesign',
+           'multinomial_qmc', 'NormalQMC', 'MultivariateNormalQMC']
+
+
+def scale(sample, bounds, reverse=False):
+    """Sample scaling from unit hypercube to bounds range.
+
+    To convert a sample from :math:`[0, 1)` to :math:`[a, b), b>a`, the
+    following transformation is used:
+
+    .. math::
+
+        (b - a) * sample + a
+
+    Parameters
+    ----------
+    sample : array_like (n_samples, dim)
+        Sample to scale.
+    bounds : tuple or array_like ([min, dim], [max, dim])
```

Comment:
That will need to be a little more clear, then. Consider accepting an array of lower bounds and an array of upper bounds separately. It a little harder for the user to make a mistake, and consequently it's a little easier to document clearly, I think. [Here](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.Bounds.html) is an example of another place in SciPy where this is the way bounds are specified. (There are others, e.g. `linprog`, that take pairs of upper and lower bounds, but after working with that code a lot, I really wish the original authors had asked for them separately!)

Reply:
ok sure I can use copy what's in `scipy.optimize` if you want. Or simply just change the shape to be `((min, max) * d)`.

Reply:
I misunderstood your comment. I did not get that you wanted to change `scale(..., bounds)` to something like `scale(..., l_bounds, u_bounds)`. What I am seeing from existing code, here it is using the *new way*. Cf. `scipy.optimize._constraints.old_bound_to_new`. So I am not sure about providing a different interfaces here. @rgommers?

Reply:
One advantage of separate `l_bounds, u_bounds` is that it can be specified as scalars. If it's `((l1, u1), (l2, u2), ...` that's harder to do. So separating makes sense to me.

Reply:
I have update the code to have `l_bounds, u_bounds`

Reply:
Thanks.